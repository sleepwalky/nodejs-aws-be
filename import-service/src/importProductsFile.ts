import { APIGatewayProxyHandler } from 'aws-lambda';
import 'source-map-support/register';
import { S3, PutObjectCommand } from '@aws-sdk/client-s3';
import { S3RequestPresigner } from '@aws-sdk/s3-request-presigner';
import { createRequest } from '@aws-sdk/util-create-request';
import { formatUrl } from '@aws-sdk/util-format-url';
import { commonHeaders } from '../../commonHeaders';

const REGION = 'eu-west-1';
const BUCKET = 'store-imported-products';
const PREFIX = 'uploaded';

export const importProductsFile: APIGatewayProxyHandler = async (
  event,
  _context
) => {
  let signedUrl;

  const name = event.queryStringParameters?.name;
  if (!name) {
    return {
      statusCode: 400,
      headers: commonHeaders,
      body: JSON.stringify(
        {
          error: 'Query parameter name is required',
        },
        null,
        2
      ),
    };
  }
  const key = `${PREFIX}/${name}`;

  // @ts-ignore
  const v3Client = new S3({ region: REGION, signatureVersion: 'v4' });

  try {
    //Create an S3RequestPresigner object
    const signer = new S3RequestPresigner({ ...v3Client.config });
    // Create request
    const request = await createRequest(
      v3Client,
      new PutObjectCommand({
        Key: key,
        Bucket: BUCKET,
      })
    );
    // Define the duration until expiration of the presigned URL
    // const expiration = new Date(Date.now() + EXPIRATION);

    // Create and format presigned URL
    signedUrl = formatUrl(
      await signer.presign(request, {
        expiresIn: 3600,
      })
    );
    console.log(`\nPutting "${key}" using signedUrl`);
  } catch (err) {
    console.log('Error creating presigned URL', err);
    return {
      statusCode: 400,
      headers: commonHeaders,
      body: JSON.stringify(
        {
          error: err,
        },
        null,
        2
      ),
    };
  }

  return {
    statusCode: 200,
    headers: commonHeaders,
    body: signedUrl,
  };
};
