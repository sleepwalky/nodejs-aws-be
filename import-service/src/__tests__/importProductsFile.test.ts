import { importProductsFile } from '../importProductsFile';
import { createRequest } from '@aws-sdk/util-create-request';

jest.mock('@aws-sdk/s3-request-presigner');
jest.mock('@aws-sdk/client-s3');
jest.mock('@aws-sdk/util-create-request');
jest.mock('@aws-sdk/util-format-url', () => ({
  formatUrl: () => 'signedUrl',
}));

describe('importProductsFile', () => {
  it('should return 400 if name is not passed in queryparams', async () => {
    const result = await importProductsFile({} as any, {} as any, () => {});
    expect(result).toMatchSnapshot();
  });

  it('should return 200 and signed url', async () => {
    const result = await importProductsFile(
      {
        queryStringParameters: {
          name: 'test',
        },
      } as any,
      {} as any,
      () => {}
    );
    expect(result).toMatchSnapshot();
  });

  it('should return 500 if something goes wrong', async () => {
    (createRequest as jest.Mock).mockRejectedValue('error');
    const result = await importProductsFile(
      {
        queryStringParameters: {
          name: 'test',
        },
      } as any,
      {} as any,
      () => {}
    );

    expect(result).toMatchSnapshot();
  });
});
