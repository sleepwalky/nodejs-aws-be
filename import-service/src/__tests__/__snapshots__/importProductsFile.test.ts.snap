// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importProductsFile should return 200 and signed url 1`] = `
Object {
  "body": "signedUrl",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;

exports[`importProductsFile should return 400 if name is not passed in queryparams 1`] = `
Object {
  "body": "{
  \\"error\\": \\"Query parameter name is required\\"
}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 400,
}
`;

exports[`importProductsFile should return 500 if something goes wrong 1`] = `
Object {
  "body": "{
  \\"error\\": \\"error\\"
}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 500,
}
`;
